### DISCUSSION TITLE
Attempt to make the output of failed command output expectations more helpful

### DISCUSSION THREAD

**Discussion by Krisell** [laravel/framework#41809] [Ideas]
Hi all,

I am trying to put together a PR to improve the output of failed `expectsOutput` or `expectsOutputContains` when testing artisan commands. However I am not completely satisfied with the PR yet and would like some help, especially with the Mockery-parts.

Please see the draft PR here, along with some questions: https://github.com/Krisell/framework/pull/2

---------------------

#### For context – here's the planned PR description
Currently, when testing artisan commands and using `expectsOutput` or `expectsOutputContains`, the error message on a failed expectation is not as helpful as it could be.

Before:
```
1) Tests\Feature\TestExpectedOutput::email_sent_test
Output "Email sent to 17 users." was not printed.
```

After:
```
1) Tests\Feature\TestExpectedOutput::email_sent_test
Output "Email sent to 17 users." was not printed.

Actual output:
"Email sent to 23 users."
```

Previously, the developer didn't know if no output was generated at all or almost the correct output was generated. If there is indeed no output, that is also mentioned with this PR:

```
1) Tests\Feature\TestExpectedOutput::email_sent_test
Output "Email sent to 17 users." was not printed.

No output was printed.
```

Of course, if there are multiple lines of output, all are displayed.
```
1) Tests\Feature\TestExpectedOutput::email_sent_test
Output "User with id 3 was updated" was not printed.

Actual output:
"Starting process ..."
"User with id 5 was updated."
"All done!"
```

**Comment by shaedrich** [comment]
That would be quite helpful 👍🏻

**Comment by shaedrich** [comment]
I actually hoped to make this even more precise
|   | Expected output                               | Actual output                                   |
| - | --------------------------------------------- | ----------------------------------------------- |
|   | Fetching user …                               | Fetching user …                                 |
| ➡ | Lorem ipsum $\textsf{\color{red}{dolor sit}}$ | Lorem ipsum $\textsf{\color{green}{ahmet sun}}$ |
|   | Done.                                         | Done.                                           |

but this might be rather complicated to achieve
